### DISCUSSION TITLE
How to upload custom Three.js code into a room?

### DISCUSSION THREAD

**Discussion by animanoir** [mozilla/hubs#2813] [General Questions & Discussion] [answered]
I'm not saying a simple 3D model, but something as sophisticated as [this example (mirrors)](https://threejs.org/examples/webgl_mirror.html) .

How could I do it?

**Comment by emclaren** [comment, marked as answer]
Hi Animanoir, 

If you are using hubs.mozilla.com (not Hubs Cloud), it is not currently possible to upload custom code into a room. 

However, if you are using Hubs Cloud (which I believe you are), I think you could do this with a custom client (https://hubs.mozilla.com/docs/hubs-cloud-custom-clients.html). By creating your own version of the client you would be able to add any of your own code in the backend. You would need to be comfortable working with code to do this, plus maintaining your custom deployment would be somewhat more work overall (we could not automatically push updates, so you would need to do these manually if you wanted to your version to stay up to date with Hubs).

**Reply by ocularvr** [reply]
This is a good answer but unfortunately the lack of assumptions behind it just make what is said not possible for a very high percentage of users. Integrating Three.js is a very needed feature and many people like myself want to know how to integrate things like this on hubs. 

We cant. There is just not enough documentation. So its a case if a really good product, ideal time, but premature release. We are grateful but at the same time have a very uneasy feeling because as in this case, man building and redeploying the whole thing on AWS is like an aircraft carrier!!!!!!

**Reply by animanoir** [reply]
That's true. I'm really enjoying learning Hubs and all the technologies underneath. But yeah, once uploading custom code is integrated, Hubs Cloud will lift off.

**Reply by ocularvr** [reply]
Im going to do the same as you, I'm sure I don't need to redeploy my stack. Simple things like doing an ftp to the server and setting mime file types would do the trick, maybe im being naive but i think that's the way to get things like the mirror up and running. Man is that a nice example!
